docs/rfcs/rfc04.md

```diff
@@ -0,0 +1,551 @@
+1. Developers have to be constantly reminded to never change the original templates. That's hard to achieve, especially when one ends up with a wall of text full of warnings in the pull request template. Developers (and reviewers) are bound to become yellow blind soon.
+1. Even a seasoned Jsonnet user will find it annoying to craft the change as an overlay and also to later merge it back in the main template files.
+1. It's hard to automate the merging of the overlay back in the original templates. This means the developer must remember to follow the rollout process and manually promote the change when it's safe to do so. That's toil and it's also brittle.
+
+## Insights
+
+1. The main insight is to recognize that the Jsonnet template **is a program**.
+
+2. The other insight is to recognize that the cluster specific entrypoints are the program. The main application template they import is just **a dependency**.
+
+   In the same vein as when you have your application A that depends on your library L at version V1, and another application B that depends on your library L at version V2, the templates for cluster A depend on the templates from library L at version V1 and cluster B depends on the templates from library L at version V2.
+
+   Most programming languages have package managers that express dependencies that include name _and_ version. If we do the same with the dependency between the cluster templates and the _application templates_ we'd solve the [problem](#Problem).
+
+3. The reason we don't normally think in terms of (1) and (2) is that the cluster-specific Jsonnet entry points are small and the bulk of the templates live in the _library_. Because of this it goes against the grain to fully split the _applications_ (cluster entrypoints like our `deploy/prod/aws/us-east-1.jsonnet`) from the _library_ (the `app/app.jsonnet`). It would be too many repositories to handle. It's too much work, and we're drawn to think of it in another way
+
+4. The last insight is that we don't need to split in separate repos. All we need is a way to refer to old _revisions_ within the same repository.
+
+Luckily Jsonnet an extensible offers import hooks functionality that kubecfg already taps extensively.
+
+## Proposal
+
+We propose to add a `srcfs` URL scheme and an import hook implementation that will allow to import file from the current project's version control system as it was a specific revision.
```

Comment:
It wasn't immediately clear to me what `srcfs` would do; I had to read your description here. 

I think you mean `src` as in "source control." If you intend to leave the implementation open to version control systems other than git, then perhaps this could be `vcsfs`?

This could be a problem that others won't share with me, so please speak out if I'm alone here.

Reply:
yeah; the reason the name is srcfs is a reference/allusion to google's internal tool; in pretty much the same way as kubecfg is also an allusion.

I think I'll move to something more explicit like `git+file:` or something `local-git:`